evolver/hardware/standard/pump.py

```diff
@@ -0,0 +1,112 @@
+        inputs = copy(self.committed)
+        inputs.update(self.proposal.items())
+        cmd = [b"--"] * self.slots
+        for pump, input in inputs.items():
+            if pump in self.ipp_pumps:
+                # TODO: calibation transform here. The IPP pumps should return array of
+                # hz values per solenoid number - containing 3 of them.
+                hz_a = [input.flow_rate] * 3
+                for solenoid, hz in enumerate(hz_a):
+                    # solenoid is 1-indexed on hardware
+                    cmd[pump * 3 + solenoid] = f"{hz}|{pump}|{solenoid+1}"
+            elif any(pump - i < 3 for i in self.ipp_pumps):
+                raise ValueError(f"pump slot {pump} reserved for IPP pump, cannot address as standard")
+            else:
+                # TODO: calibration transform here. The time_to_pump is the target variable
+                # where pump_interval would presumably be what the pump_interval was set to
+                # during calibration and is fixed post-calibration.
+                time_to_pump, pump_interval = (input.flow_rate, int(input.flow_rate))
+                cmd[pump] = f"{time_to_pump}|{pump_interval}".encode()
+        with self.serial as comm:
+            comm.communicate(SerialData(addr=self.addr, data=cmd))
+        self.committed = inputs
+
+
+class VialIEPump(EffectorDriver):
+    """Vial-based Influx-Efflux Pump driver.
+
+    Each pump in the array is either an influx (to bring liquid to the vial) or
+    efflux (to pull waste out of vial) or spare (unused). Flow rates can be configured
+    separately or as a single rate.
+
+    Configuration parameters ``influx_map`` and ``efflux_map`` can be used to specify
+    explicit mapping between vial and underlying pump_id (of :py:class:`GenericPump`),
+    which by default reserves the first 3rd of slots to influx, and second third to
+    efflux (remaining are spare).
+    """
+
+    class Config(GenericPump.Config):
+        influx_map: dict[int, int] | None = Field(None, description="map of vial to influx pump ID")
+        efflux_map: dict[int, int] | None = Field(None, description="map of vial to efflux pump ID")
+
+    class Input(VialBaseModel):
+        flow_rate_influx: float = Field(description="influx flow rate in ml/s")
+        flow_rate_efflux: float | None = Field(None, description="efflux flow rate in ml/s. Defaults to same as influx")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        p_slots = self.slots * 3  # influx, efflux, spare (or 3 solenoids for IPP)
+        self._generic_pump = GenericPump(
+            addr=self.addr,
+            slots=p_slots,
+            serial=self.serial_conn,
+            ipp_pumps=self.ipp_pumps,
+            evolver=kwargs.get("evolver"),
+        )
+        self.influx_map = self.influx_map or {i: i for i in range(0, p_slots)}
+        self.efflux_map = self.efflux_map or {i: i + self.slots for i in range(0, p_slots)}
```

Comment:
FYI this kind of default inter-field initialization can happen in [model_post_init](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_post_init).

We should discuss whether this is something we would prefer or even need, e.g., does the config need to be complete & correct without the need to instantiate its container?  Would this be needed for the UI, to render the efflux map from the config alone?

Reply:
The `model_post_init` is that of `Config` if I understand you correctly. Yeah I can add it there. Could you elaborate on the second part? Here I take None to represent that we want the default arrangement so a user doesn't have to explicitly map - the UI could likewise indicate that no setting was explicitly given by the user.

Reply:
Do you think we may need an endpoint to create model instances for the UI to hit and have the complete model instance returned? If so, it would be better for us to be able to instantiate a ``Config`` model instance and have it be complete with defaults and not have to instantiate the container class to achieve this. In some scenarios we really may not want to instantiate the container class due to other unwanted side-effects, e.g., opening device connections etc. This aspect is also just true for SDK "correctness" - the ``Config`` model should be responsible for its default, not the container (at least were possible).

Reply:
Well, this is just the exact same thing as above but from the other side, you equally can't instantiate a ``Config`` because it wouldn't be processing the default case.

However, we've discussed this pattern in the past and you decided that you wanted to not duplicate param semantics, defaults, and init functionality across both the container and its ``Config``. The solution is to use the creation pattern we settled on in #73 ( I think that's the correct PR) and call ``class.create()`` which first instantiates a ``Config`` instance and then passes that to ``class.__init__``.

Reply:
Specifically I mean here: https://github.com/ssec-jhu/evolver-ng/blob/main/evolver/hardware/test_utils.py#L117 in the tests we instantiate through several available means. If I move the default setting out to Config, the one at that line fails (but I can indeed instantiate a Config since `None` is an allowed value, which is represented in another test).

I'm not sure what to do here - shall I do the repetition of default settings, or get rid of the direct instantiation test (and thereby assume that these should be created with `create` instead)?

Reply:
I'd go with the latter and use ``create``.

That being said, there's already enough complexity in [evolver.base.BaseInterface.auto_assign_attrs_from_config](https://github.com/ssec-jhu/evolver-ng/blob/main/evolver/base.py#L266) that there could be some argument made for seeing if ``Config.model_post_init`` can also be accommodated. On the other hand, us having to incorporate/replicate ``pydantic`` functionality was my main reason for not wanting to add ``auto_assign_attrs_from_config`` in the first place - so 🤷